tensorboard/data/server/commit.rs

```diff
@@ -136,3 +136,166 @@ mod tests {
+        /// time if unset.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use rustboard_core::commit::{test_data::CommitBuilder, Commit};
+        /// use rustboard_core::types::Step;
+        ///
+        /// let my_commit: Commit = CommitBuilder::new()
+        ///     .scalars("train", "xent", |mut b| {
+        ///         b.eval(|Step(i)| 1.0 / (i + 1) as f32).build()
+        ///     })
+        ///     .build();
+        /// ```
+        pub fn scalars(
+            self,
+            run: &str,
+            tag: &str,
+            build: impl FnOnce(ScalarTimeSeriesBuilder) -> TimeSeries<ScalarValue>,
+        ) -> Self {
+            self.with_run_data(Run(run.to_string()), |run_data| {
+                let time_series = build(ScalarTimeSeriesBuilder::default());
+                if let (None, Some((_step, wall_time, _value))) =
+                    (run_data.start_time, time_series.valid_values().next())
+                {
+                    run_data.start_time = Some(wall_time);
+                }
+                run_data.scalars.insert(Tag(tag.to_string()), time_series);
+            });
+            self
+        }
+
+        /// Ensures that a run is present and sets its start time.
+        ///
+        /// If you don't care about the start time and the run is going to have data, anyway, you
+        /// can just add the data directly, which will create the run as a side effect.
+        ///
+        /// # Panics
+        ///
+        /// If `start_time` represents an invalid wall time (i.e., `start_time` is `Some(wt)` but
+        /// `WallTime::new(wt)` is `None`).
+        pub fn run(self, run: &str, start_time: Option<f64>) -> Self {
+            self.with_run_data(Run(run.to_string()), |run_data| {
+                run_data.start_time = start_time.map(|f| WallTime::new(f).unwrap());
+            });
+            self
+        }
+
+        /// Consumes this `CommitBuilder` and returns the underlying commit.
+        pub fn build(self) -> Commit {
+            self.0
+        }
+    }
+
+    pub struct ScalarTimeSeriesBuilder {
+        /// Initial step. Increments by `1` for each point.
+        step_start: Step,
+        /// Initial wall time. Increments by `1.0` for each point.
+        wall_time_start: WallTime,
+        /// Number of points in this time series.
+        len: u64,
+        /// Custom summary metadata. Leave `None` to use default.
+        metadata: Option<Box<pb::SummaryMetadata>>,
+        /// Scalar evaluation function, called for each point in the series.
+        ///
+        /// By default, this maps step `i` to `0.5 ** (i + 1)`.
+        eval: Box<dyn Fn(Step) -> f32>,
+    }
+
+    impl Default for ScalarTimeSeriesBuilder {
+        fn default() -> Self {
+            ScalarTimeSeriesBuilder {
+                step_start: Step(0),
+                wall_time_start: WallTime::new(1235.0).unwrap(),
```

Comment:
Optional, but left to my own devices I would probably either omit some of these default values (aka require them to be set on the builder) or at least try to have the defaults be sort of the "minimally arbitrary" values (e.g. start wall time at 0.0 incrementing by 1.0, make len 1, make `eval` just be 0 always, etc.).

Otherwise we still have `test_read_scalars()` making assertions like `assert_eq!(xent_data.wall_time, vec![1235.0, 1236.0, 1237.0]);` that can't be understood to be correct without consulting this default implementation.  By contrast, `test_list_scalars()` is explicit at least about the `wall_time_start` and `len` values and thus its assertions are self-contained.

Reply:
Thanks. You have inspired me to be less lazy. Done.

Reply:
`<meme>what is my purpose?</meme>`